File: content-solutions/dfs/orchard-split.java
import java.util.*;

class Solution {
    List<List<Integer>> graph = new ArrayList<>();
    int[] apples;
    int total, best;

    public int solution(int[] apples, int[][] roads) {
        this.apples = apples;
        int n = apples.length;
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        for (int[] r : roads) {
            graph.get(r[0]).add(r[1]);
            graph.get(r[1]).add(r[0]);
        }
        for (int a : apples) total += a;
        best = total;
        dfs(0, -1);
        return best;
    }

    int dfs(int u, int parent) {
        int s = apples[u];
        for (int v : graph.get(u)) {
            if (v != parent) s += dfs(v, u);
        }
        if (parent != -1) best = Math.min(best, Math.abs(total - 2 * s));
        return s;
    }
}

File: content-solutions/dfs/orchard-split.js
function solution(apples, roads) {
  const n = apples.length;
  const graph = Array.from({ length: n }, () => []);
  for (const [a, b] of roads) {
    graph[a].push(b);
    graph[b].push(a);
  }
  const total = apples.reduce((x, y) => x + y, 0);
  let best = total;
  function dfs(u, parent) {
    let s = apples[u];
    for (const v of graph[u]) {
      if (v !== parent) s += dfs(v, u);
    }
    if (parent !== -1) best = Math.min(best, Math.abs(total - 2 * s));
    return s;
  }
  dfs(0, -1);
  return best;
}
